### PR TITLE
add embedding support in main.py

### LINKED ISSUE #14: feat: Support embedding API
Support this API: https://platform.openai.com/docs/api-reference/embeddings/create

Several endpoints:

- [ ] /engines/*/embeddings
- [ ] /v1/engines/*/embeddings
- [ ] /embeddings
- [ ] /v1/embeddings

### PR DESCRIPTION
**Pull request by tddschn**
Resolves #14

### CODE REVIEW

**Review by kemingy — CHANGES_REQUESTED** [review]

File: main.py [hunk 1]
```diff
@@ -153,14 +156,44 @@ async def on_post(self, req: Request, resp: Response):
...
 app = App()
 app.add_route("/", Ping())
 app.add_route("/completions", Completions(model_name=MODEL))
 app.add_route("/chat/completions", ChatCompletions(model_name=MODEL))
+app.add_route("/embeddings", Embeddings(model_name=EMBEDDING_MODEL))
 # refer to https://platform.openai.com/docs/api-reference/chat
 # make it fully compatible with the current OpenAI API endpoints
 app.add_route("/v1/completions", Completions(model_name=MODEL))
 app.add_route("/v1/chat/completions", ChatCompletions(model_name=MODEL))
+app.add_route("/v1/embeddings", Embeddings(model_name=EMBEDDING_MODEL))
```
Comment (gaocegege) [inline main.py:196]: Also need to support engines/<>/embeddings

Comment (kemingy) [reply main.py:213]: Any reference?

Comment (gaocegege) [reply main.py:211]: From the OpenAI python package implementation:

https://github.com/openai/openai-python/blob/main/openai/api_resources/abstract/engine_api_resource.py

File: main.py [hunk 2]
```diff
@@ -163,37 +166,55 @@ def __init__(self, model_name: str) -> None:
...
+    "/engines/{}/embeddings".format(EMBEDDING_MODEL),
+    Embeddings(model_name=EMBEDDING_MODEL),
+)
 # refer to https://platform.openai.com/docs/api-reference/chat
 # make it fully compatible with the current OpenAI API endpoints
 app.add_route("/v1/completions", Completions(model_name=MODEL))
 app.add_route("/v1/chat/completions", ChatCompletions(model_name=MODEL))
 app.add_route("/v1/embeddings", Embeddings(model_name=EMBEDDING_MODEL))
+app.add_route(
+    "/v1/engines/{}/embeddings".format(EMBEDDING_MODEL),
```
Comment (gaocegege) [inline main.py:213]: Could we make engine name a wildcard? Because we do not care about the engine name actually.

Comment (tddschn) [reply main.py:213]: @kemingy Is this correct in falcon?

Comment (gaocegege) [reply main.py:213]: https://github.com/falconry/falcon/issues/648

File: main.py [hunk 3]
```diff
@@ -18,12 +18,28 @@
...
 )
+
+
+# todo: make this importable from top level
+# from llmspec.llmspec import EmbeddingData
+# temporary fix: embedding attr type
+class EmbeddingData(msgspec.Struct):
+    embedding: List[float] | List[List[float]]
+    index: int
+    object: str = "embedding"
```
Comment (kemingy) [inline main.py:32]: You can use the latest llmspec code from your branch. Just mark this PR as blocked by another one.

File: main.py [hunk 4]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
+
+    async def on_post(self, req: Request, resp: Response):
+        buf = await req.stream.readall()
+        try:
+            # todo: llmspec hasn't implemented from_bytes for EmbeddingRequest
+            embedding_req = msgspec.json.decode(buf, type=EmbeddingRequest)
+        except msgspec.ValidationError as err:
+            logger.info(f"Failed to parse request: {err}")
+            resp.status = falcon.HTTP_400
+            resp.status = falcon.HTTP_400
```
Comment (kemingy) [inline main.py:184]:
```suggestion
```

File: main.py [hunk 5]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
 
+class Embeddings:
+    def __init__(self, model_name: str) -> None:
+        self.model_name = model_name
+
+    async def on_post(self, req: Request, resp: Response):
+        buf = await req.stream.readall()
+        try:
+            # todo: llmspec hasn't implemented from_bytes for EmbeddingRequest
+            embedding_req = msgspec.json.decode(buf, type=EmbeddingRequest)
```
Comment (kemingy) [inline main.py:180]: Why not impl it in llmspec?

File: main.py [hunk 6]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
+        except msgspec.ValidationError as err:
+            logger.info(f"Failed to parse request: {err}")
+            resp.status = falcon.HTTP_400
+            resp.status = falcon.HTTP_400
+            resp.data = ErrorResponse.from_validation_err(err, str(buf)).to_json()
+            return
+
+        model = SentenceTransformer(self.model_name)
```
Comment (kemingy) [inline main.py:188]: DO NOT load model every time.

File: main.py [hunk 7]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
+            embeddings = embeddings.tolist()
+        else:
+            embeddings = embeddings.tolist()
+
+        embedding_data = EmbeddingData(embedding=embeddings, index=0)
+        embedding_resp = EmbeddingResponse(
+            data=embedding_data,
+            model=self.model_name,
+            usage=TokenUsage(
+                prompt_tokens=0,  # No prompt tokens, only embeddings generated.
```
Comment (kemingy) [inline main.py:203]: https://platform.openai.com/docs/api-reference/embeddings/create

File: main.py [hunk 8]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
+                completion_tokens=0,  # No completions performed, only embeddings generated.
+                total_tokens=len(embeddings),
+            ),
+        )
+        # todo: llmspec hasn't implemented to_json for EmbeddingResponse
+        resp.data = msgspec.json.encode(embedding_resp)
```
Comment (kemingy) [inline main.py:209]: Impl it.

File: main.py [hunk 9]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
+
+
+class EmbeddingsEngineRouteWrapper(Embeddings):
+    def __init__(self, model_name: str) -> None:
+        super().__init__(model_name)
+
+    async def on_post(self, req: Request, resp: Response, engine: str):
+        await super().on_post(req, resp)
```
Comment (kemingy) [inline main.py:217]: Seems it equals to the parent class?

File: main.py [hunk 10]
```diff
@@ -153,14 +169,72 @@ async def on_post(self, req: Request, resp: Response):
...
+
+
 app = App()
 app.add_route("/", Ping())
 app.add_route("/completions", Completions(model_name=MODEL))
 app.add_route("/chat/completions", ChatCompletions(model_name=MODEL))
+app.add_route("/embeddings", Embeddings(model_name=EMBEDDING_MODEL))
```
Comment (kemingy) [inline main.py:224]: DO NOT load the model for 4 times.

**Review by kemingy — CHANGES_REQUESTED** [review]

File: main.py (hunk 2 above)
Comment (kemingy) [reply main.py]: https://falcon.readthedocs.io/en/stable/api/routing.html

Of course every HTTP web framework supports this.

File: main.py (hunk 9 above)
Comment (tddschn) [reply main.py:216]: parent class doesn't accept the engine arg. Do you want me to add the arg to Embeddings?

File: main.py [hunk 11]
```diff
@@ -153,14 +161,110 @@ async def on_post(self, req: Request, resp: Response):
         resp.data = completion.to_json()
 
 
+class Embeddings:
+    def __init__(self, model_name: str) -> None:
+        self.model_name = model_name
+        self.model = SentenceTransformer(self.model_name)
```
Comment (kemingy) [inline main.py:167]: Undefined?

File: main.py [hunk 12]
```diff
@@ -153,14 +161,110 @@ async def on_post(self, req: Request, resp: Response):
...
+        from transformers import AutoTokenizer, AutoModel
```
Comment (kemingy) [inline main.py:168]: `import` should be placed on the top of the file.

File: main.py [hunk 13]
```diff
@@ -153,14 +161,110 @@ async def on_post(self, req: Request, resp: Response):
...
+
+        # Load model from HuggingFace Hub
+        self.tokenizer = AutoTokenizer.from_pretrained(self.model)
+        self.model = AutoModel.from_pretrained(self.model)
```
Comment (kemingy) [inline main.py:172]: Why not use sentence_transformers

File: main.py [hunk 14]
```diff
@@ -153,14 +161,110 @@ async def on_post(self, req: Request, resp: Response):
...
+        with torch.no_grad():
+            model_output = self.model(**encoded_input)
+
+        # Perform pooling
+        sentence_embeddings = mean_pooling(
+            model_output, encoded_input['attention_mask']
+        )
+
+        # Normalize embeddings
+        sentence_embeddings = F.normalize(sentence_embeddings, p=2, dim=1)
```
Comment (kemingy) [inline main.py:203]: Is it copied from the sentence_transformers? If it doesn't expose this interface, you can skip the token count for now and create another issue to track it.

File: main.py [hunk 15]
```diff
@@ -153,14 +161,110 @@ async def on_post(self, req: Request, resp: Response):
...
+class EmbeddingsEngineRouteWrapper(Embeddings):
+    def __init__(self, model_name: str) -> None:
+        super().__init__(model_name)
+
+    async def on_post(self, req: Request, resp: Response, engine: str):
+        await super().on_post(req, resp)
+
+
+embeddings = Embeddings(EMBEDDING_MODEL)
+embeddings_engine_route_wrapper = EmbeddingsEngineRouteWrapper(EMBEDDING_MODEL)
```
Comment (kemingy) [inline main.py:248]: This still loads two models into the memory.

**Comment by tddschn** [comment]
how do I mark this PR as blocked by https://github.com/tensorchord/llmspec/pull/27 ?

https://stackoverflow.com/questions/26619478/are-dependent-pull-requests-in-github-possible

I don't think github support this at this time.

**Comment by kemingy** [comment]
Can you resolve the conflicts?

**Comment by kemingy** [comment]
Merging this since it blocks the progress. Will refactor it later.

**Pull request merged**